### PR TITLE
Add OscArray type

### PR DESCRIPTION
**Pull request by bitzl**
Complete implementation for OSC array types. Arrays in this implementation may contain elements with different types, and all types are supported. This includes other arrays, which can be nested as needed.

~~Experimental encoding implementation for OSC array types to see how this could look like. I'm sure there's still room for improvement, so it's a draft for now.~~

### CODE REVIEW

File: src/encoder.rs [hunk 1]
```diff
@@ -1,7 +1,7 @@
-use types::{Result, OscType, OscPacket, OscBundle, OscMessage};
 use errors::OscError;
+use types::{OscBundle, OscMessage, OscPacket, OscType, Result};
 
-use byteorder::{ByteOrder, BigEndian};
```
Comment (klingtnet) [inline src/encoder.rs]: If it's not too much hassle please try to omit the formatting changes or put them in a separate commit. This makes reviewing a bit easier.

File: src/encoder.rs [hunk 2]
```diff
@@ -91,34 +91,34 @@ fn encode_bundle(bundle: &OscBundle) -> Result<Vec<u8>> {
...
 }
 
-fn encode_arg(arg: &OscType) -> Result<(Option<Vec<u8>>, char)> {
+fn encode_arg(arg: &OscType) -> Result<(Option<Vec<u8>>, String)> {
     match *arg {
         OscType::Int(ref x) => {
             let mut bytes = vec![0u8; 4];
             BigEndian::write_i32(&mut bytes, *x);
-            Ok((Some(bytes), 'i'))
+            Ok((Some(bytes), 'i'.to_string()))
```
Comment (klingtnet) [inline src/encoder.rs]: You can replace the `.to_string()` calls by wrapping the chars in double quotes.

Comment (klingtnet) [reply src/encoder.rs]: Ok, it's not that easy but `"x".into()` works.

File: src/encoder.rs [hunk 3]
```diff
@@ -127,14 +127,40 @@ fn encode_arg(arg: &OscType) -> Result<(Option<Vec<u8>>, char)> {
...
-            Ok((Some(bytes), 'b'))
+            Ok((Some(bytes), 'b'.to_string()))
         }
-        OscType::Time(ref x, ref y) => Ok((Some(encode_time_tag(*x, *y)), 't')),
-        OscType::Midi(ref x) => Ok((Some(vec![x.port, x.status, x.data1, x.data2]), 'm')),
-        OscType::Color(ref x) => Ok((Some(vec![x.red, x.green, x.blue, x.alpha]), 'r')),
-        OscType::Bool(ref x) => if *x { Ok((None, 'T')) } else { Ok((None, 'F')) },
-        OscType::Nil => Ok((None, 'N')),
-        OscType::Inf => Ok((None, 'I')),
+        OscType::Array(ref x) => {
```
Comment (klingtnet) [inline src/encoder.rs]: I would like to move the `Array` case to the bottom since its the most complex one.

File: src/encoder.rs [hunk 4]
```diff
@@ -127,14 +127,40 @@ fn encode_arg(arg: &OscType) -> Result<(Option<Vec<u8>>, char)> {
...
+            let mut bytes = vec![0u8; x.len()];
+            let mut type_tags = String::from("[");
+            for v in x.iter() {
+                match encode_arg(v).unwrap() {
```
Comment (klingtnet) [inline src/encoder.rs]: I know it's a draft but we must handle the error here. `match encode_arg(v)?` should work.

File: src/encoder.rs [hunk 5]
```diff
@@ -127,14 +127,40 @@ fn encode_arg(arg: &OscType) -> Result<(Option<Vec<u8>>, char)> {
...
+                    (Some(other_bytes), other_type_tags) => {
+                        bytes.extend(other_bytes);
+                        type_tags.extend(other_type_tags.chars());
```
Comment (klingtnet) [inline src/encoder.rs]: Not sure if this more idiomatic but `String`s provide `push`: `type_tags.push_str(&other_type_tags);`.

File: src/encoder.rs [hunk 6]
```diff
@@ -127,14 +127,40 @@ fn encode_arg(arg: &OscType) -> Result<(Option<Vec<u8>>, char)> {
...
+                    }
+                    (None, other_type_tags) => {
+                        type_tags.extend(other_type_tags.chars());
+                    }
+                }
+            }
+            type_tags = type_tags + "]";
```
Comment (klingtnet) [inline src/encoder.rs]: We can use (for consistency) `type_tags.push(']');` here.

File: src/encoder.rs (hunk 1 above)
Comment (bitzl) [reply src/encoder.rs]: Thank you for pointing this out. Seems like there is some auto-formatting going on that I can't turn off (VS Code  and rustfmt?), so I'll be careful when adding changes to Git.

**Comment by klingtnet** [comment]
You may have to rebase from since I did something stupid and force pushed. Nonetheless, it should be the same state as before.

**Comment by bitzl** [comment]
> You may have to rebase from `master` since I did something stupid and force pushed. Nonetheless, it should be the same state as before. Sorry for the inconvenience.
> Also, please ask me if there's something I can help you with implementing the feature.

Actually, I don't see any problems. Maybe GitHub somehow automagically rebased my fork? Perhaps you could have a look and tell me if there's something I'm missing.

The array type implementation should be complete now, including encoding, decoding and tests for both. Please let me know if there's anything I can (or should) improve.

**Comment by klingtnet** [comment]
Thank you for the update!

> Actually, I don't see any problems. Maybe GitHub somehow automagically rebased my fork? Perhaps you could have a look and tell me if there's something I'm missing.

I was able to force push the previous state again.

> The array type implementation should be complete now, including encoding, decoding and tests for both. Please let me know if there's anything I can (or should) improve.

I greatly appreciate your efforts and will take a look on it somewhen this evening. :+1:

**Comment by klingtnet** [comment]
Thnaks again! Will merge and release!

**Comment by ClosetMonkey** [comment]
I noticed you allow arrays of arrays (nested arrays). I’m struggling to find if this is a normal feature for OSC implementations. The spec doesn’t seem to explicitly say either way.
